04_matrices/ref-craters.py

```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3
+
+import sys
+
+
+"""
+    Inits a matrix of size lines x cols with all slots set at value
+"""
+def init(lines, cols, value):
+
+    visitedMap = []
+
+    for i in range(lines):
+
+        L = []
+
+        for j in range(cols):
+            L.append(value)
+
+        visitedMap.append(L)
+
+    return visitedMap
+
+
+"""
+    Creates a matrix from a txt file
+"""
+def loadMap(filename):
+
+    M = []
+    f = open(filename)
+    lines = f.readlines()
+    f.close()
+
+    for l in lines:
+        M.append(l)
+
+    return M;
+
+
+"""
+    Visits the whole crater
+"""
+def replaceCrater(M, visitedMap, i, j):
+
+    visitedMap[i][j] = True
+
+    for k in range(-1, 2):
+        for l in range(-1, 2):
+            if M[i + k][j + l] == '#' and visitedMap[i + k][j + l] == False:
```

Comment:
Si (i, j) est sur un bord de la matrice, tu vas faire un accès out of bounds.
Tu peux virer ce if, le mettre au début de la fonction, et rajouter une vérification que (i, j) est bien dans la matrice.
```python
def replaceCrater(M, visitedMap, i, j):
    if i < 0 or i > len(M):
        return
    if j < 0 or j > len(M[0]):
        return
    if M[i][j] == '#' or visitedMap[i][j] == True:
        return
    # processing 
```